src/script/view_model/CallingViewModel.ts

```diff
@@ -148,23 +156,138 @@ export class CallingViewModel {
+          },
+        );
+
+        callingSubscriptions.addOngoing(call.conversationId, unsubscribe);
+      }
       ring(call);
     };
 
+    const generateSubconversationMembers = async (
+      conversationId: QualifiedId,
+      subconversation: Subconversation,
+    ): Promise<SubconversationEpochInfoMember[]> => {
+      const mlsService = core.service?.mls;
+
+      if (!mlsService) {
+        throw new Error('mls service was not initialised');
+      }
+
+      const parentGroupId = await mlsService.getGroupIdFromConversationId(conversationId);
+      const memberIds = await mlsService.getClientIds(parentGroupId);
+
+      const members = memberIds.map(parentMember => {
+        const isSubconversationMember = !!subconversation.members.find(
+          ({user_id, client_id}) => user_id === parentMember.userId && client_id === parentMember.clientId,
+        );
+
+        return {
+          userid: `${parentMember.userId}@${parentMember.domain}`,
+          clientid: parentMember.clientId,
+          in_subconv: isSubconversationMember,
+        };
+      });
+
+      return members;
+    };
+
+    const getSubconversationEpochInfo = async (
+      conversationId: QualifiedId,
+      subconversation: Subconversation,
+    ): Promise<{
+      members: SubconversationEpochInfoMember[];
+      epoch: number;
+      secretKey: string;
+      keyLength: number;
+    }> => {
+      const mlsService = core.service?.mls;
+      if (!mlsService) {
+        throw new Error('mls service was not initialised');
+      }
+
+      const members = await generateSubconversationMembers(conversationId, subconversation);
+
+      const epoch = Number(await mlsService.getEpoch(subconversation.group_id));
+
+      const keyLength = 32;
+      const secretKey = await mlsService.exportSecretKey(subconversation.group_id, keyLength);
+
+      return {members, epoch, keyLength, secretKey};
+    };
+
+    const subscribeToEpochUpdates = async (
+      conversationId: QualifiedId,
+      onEpochUpdate: (info: {
+        members: SubconversationEpochInfoMember[];
+        epoch: number;
+        secretKey: string;
+        keyLength: number;
+      }) => void,
+    ): Promise<() => void> => {
+      const mlsService = core.service?.mls;
+      if (!mlsService) {
+        throw new Error('mls service was not initialised');
+      }
+      const initialSubconversation = await mlsService.joinConferenceSubconversation(conversationId);
+
+      const forwardNewEpoch = async ({groupId, epoch}: {groupId: string; epoch: number}) => {
+        const subconversation = await mlsService.getConferenceSubconversation(conversationId);
+
+        if (groupId !== subconversation.group_id) {
+          return;
+        }
+
+        const {keyLength, secretKey, members} = await getSubconversationEpochInfo(conversationId, subconversation);
+        onEpochUpdate({epoch: Number(epoch), keyLength, secretKey, members});
+      };
```

Comment:
This is pretty important change - we need to make sure we fetch fresh subconversation every time we receive epoch change event to get the fresh list of its members. Previously we were using closed value of subconversation fetched above `forwardNewEpoch` function definition.